File: guides/openai-assistant-js/actions.ts
import { Sandbox } from '@e2b/sdk'
import path from 'path'

export async function saveCodeToFile(
  sandbox: Sandbox,
  { code, filename }: { code: string; filename: string },
): Promise<string> {
  try {
    const dir = path.dirname(filename)

    await sandbox.filesystem.makeDir(dir)
    await sandbox.filesystem.write(filename, code)

    return 'success'
  } catch (e) {
    return `Error: ${e.message}}`
  }
}

export async function listFiles(sandbox: Sandbox, { path }: { path: string }): Promise<string> {
  try {
    const files = await sandbox.filesystem.list(path)
    const response = files.map(file => (file.isDir ? `dir: ${file.name}` : file.name)).join('\n')
    return response
  } catch (e) {
    return `Error: ${e.message}}`
  }
}

export async function readFile(sandbox: Sandbox, { path }: { path: string }): Promise<string> {
  try {
    return await sandbox.filesystem.read(path)
  } catch (e) {
    return `Error: ${e.message}}`
  }
}

File: guides/openai-with-sandbox/actions.py
import os
import e2b
import openai
from e2b import Sandbox
from typing import Any, Dict # Optional - import types, allowing for explicit type hints
from dotenv import load_dotenv

load_dotenv()
openai.api_key = os.environ["OPENAI_API_KEY"]

#spawn E2B sandbox
client = openai.Client()
sandbox = Sandbox()


# Define assistant's actions
def save_code_to_file(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    filename = args["filename"]
    code = args["code"]

    try:
        dir = os.path.dirname(filename)

        sandbox.filesystem.make_dir(dir)
        sandbox.filesystem.write(filename, code)

        return "success"
    except Exception as e:
        return f"Error: {e}"


def list_files(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    path = args["path"]

    try:
        files = sandbox.filesystem.list(path)
        response = "\n".join(
            [f"dir: {file.name}" if file.is_dir else file.name for file in files]
        )
        return response
    except Exception as e:
        return f"Error: {e}"


def read_file(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    path = args["path"]

    try:
        return sandbox.filesystem.read(path)
    except Exception as e:
        return f"Error: {e}"
    
